File: frontend/src/types/kiteEngine.ts
// Types for the Kite-exclusive triple-SuperTrend options engine.
// Mirrors backend app/engines/triple_supertrend/schemas.py.

export type TrailTarget = 'fast' | 'mid' | 'slow';
export type Moneyness = 'ATM' | 'ITM1' | 'ITM2' | 'ITM3' | 'ITM4' | 'ITM5' | 'ITM10' | 'ITM15' | 'ITM20' | 'OTM1' | 'OTM2' | 'OTM3' | 'OTM4' | 'OTM5';
export type ScanSource = 'spot' | 'derivatives' | 'both';
export type ScanExpiry = 'weekly' | 'monthly';
export type Vehicle = 'otm_options' | 'deep_itm_options' | 'futures';
export type DeepItmMoneyness = 'ITM5' | 'ITM10' | 'ITM15' | 'ITM20';

export interface AlignmentChip {
  fast: number; // +1 / -1 / 0
  mid: number;
  slow: number;
}

export interface OptionLeg {
  moneyness: string; // ATM / ITM1 / ITM2 / OTM1 / OTM2
  option_type: string; // CE / PE
  option_symbol: string;
  strike: number;
  expiry: string;
  lot_size: number | null;
  premium_spot?: number;
  premium_sl?: number;
  token?: number;
  is_active?: boolean; // this contract's SuperTrend still aligned on the latest bar
}

export interface EngineSignalRow {
  underlying: string;
  token: number;
  exchange: string; // option exchange (NFO / BFO)
  regime: 'BULL' | 'BEAR';
  alignment: AlignmentChip;
  direction: 'long' | 'short';
  option_type: 'CE' | 'PE';
  legs: OptionLeg[];
  spot: number;
  stop_loss: number;
  score: number;
  timestamp_ms: number;
  source?: 'spot' | 'derivatives';
  is_active?: boolean; // SuperTrend still aligned on the latest bar (trade running)
  is_fresh?: boolean;  // entered on the latest closed bar (the live "ready now" trigger)
  adx?: number | null;      // ADX at signal time (trend strength, 0–100)
  atr_pct?: number | null;  // ATR percentile at signal time (volatility rank)
}

export interface SignalsResponse {
  generated_ms: number;
  scanning: boolean;
  scanning_label: string;
  rows: EngineSignalRow[];
  next_scan_ms: number;
  auto_scan: boolean;
  market_open: boolean;
}

export interface ActivityEvent {
  ts_ms: number;
  kind: string; // scan_start | scan_done | order_placed | order_blocked | order_failed | error | info
  message: string;
}

export interface ActivityResponse {
  events: ActivityEvent[];
  scanning: boolean;
  auto_scan: boolean;
  last_scan_ms: number;
  next_scan_ms: number;
  signal_count: number;
  scanning_label: string;
}

export interface SetupPoint {
  time: number; // epoch seconds
  open: number;
  high: number;
  low: number;
  close: number;
}

export interface SetupLine {
  time: number;
  value: number;
}

export interface SetupChart {
  underlying: string;
  candles: SetupPoint[];
  st_fast: SetupLine[];
  st_mid: SetupLine[];
  st_slow: SetupLine[];
  entry_index: number | null;
  trail_target: string;
}

export interface DepthLevel {
  price: number;
  quantity: number;
  orders: number;
}

export interface OptionDetail {
  moneyness: string;
  option_type: string;
  option_symbol: string;
  strike: number;
  expiry: string;
  lot_size: number | null;
  dte: number;
  last_price: number;
  bid: number;
  ask: number;
  iv: number; // decimal
  delta: number;
  gamma: number;
  theta: number;
  vega: number;
  depth_buy: DepthLevel[];
  depth_sell: DepthLevel[];
}

export interface EngineDetailResponse {
  underlying: string;
  token: number;
  exchange: string;
  direction: 'long' | 'short';
  regime: 'BULL' | 'BEAR';
  alignment: AlignmentChip;
  option_type: 'CE' | 'PE';
  triggered_ms: number;
  spot_at_trigger: number;
  spot_now: number;
  stop_loss: number;
  options: OptionDetail[];
}

export interface EngineConfigModel {
  engine_enabled: boolean;
  trail_target: TrailTarget;
  strike_moneyness: Moneyness[];
  scan_source: ScanSource;
  scan_expiries: ScanExpiry[];
  scan_expiries_indices?: ScanExpiry[] | null;
  scan_expiries_stocks?: ScanExpiry[] | null;
  scan_indices: string[];
  scan_stocks: string[];
  scan_all_stocks: boolean;
  early_lock: boolean;
  auto_execute: boolean;
  // Per-trade risk sizing (workstream F)
  risk_sizing: boolean;
  risk_pct: number;
  max_lots: number;
  // Protective stop mode (workstreams C/D)
  stop_mode: 'broker' | 'monitor' | 'both';
  // ── Directional mode (additive, opt-in) ────────────────────────────────
  directional_mode: boolean;
  vehicle: Vehicle;
  enabled_vehicles: Vehicle[];
  itm_depth: DeepItmMoneyness | null;
  target_delta: number | null;
  futures_expiry: 'near' | 'next';
  adx_min: number | null;
  atr_pct_min: number | null;
  wire_risk_infra: boolean;
}

// ─── Options backtest (workstream H) ─────────────────────────────────────────
export type BacktestDataMode = 'synthetic' | 'real' | 'both';

export interface BacktestRequest {
  symbol: string;
  data_mode: BacktestDataMode;
  trail_target: TrailTarget;
  lookback_bars: number;
  starting_capital: number;
  qty: number;
  iv: number;
  dte_days: number;
  moneyness_offset_pct: number;
  slippage_pct?: number | null;
  brokerage_per_order?: number | null;
}

export interface BacktestTrade {
  entry_ms: number;
  exit_ms: number;
  direction: string;
  entry_premium: number;
  exit_premium: number;
  qty: number;
  gross_pnl: number;
  costs: number;
  net_pnl: number;
  bars_held: number;
  exit_reason: string;
}

export interface BacktestStats {
  trades: number;
  wins: number;
  losses: number;
  win_rate: number;
  gross_pnl: number;
  total_costs: number;
  net_pnl: number;
  profit_factor: number;
  expectancy: number;
  avg_win: number;
  avg_loss: number;
  max_drawdown: number;
  sharpe: number;
  return_pct: number;
  final_capital: number;
}

export interface BacktestRun {
  mode: string;
  caveat: string;
  trades: BacktestTrade[];
  equity_curve: number[];
  stats: BacktestStats;
}

export interface BacktestResponse {
  symbol: string;
  data_mode: string;
  generated_ms: number;
  runs: BacktestRun[];
  bs_vs_real_drift_pct?: number | null;
  notes: string[];
}

export interface EngineOrderRequest {
  option_symbol: string;
  exchange: string;
  side: 'BUY' | 'SELL';
  quantity: number;
  order_type?: string;
  product?: string;
}

export interface EngineOrderResponse {
  order_id: string;
  status: string;
  message: string;
}

// ─── Stock registry ────────────────────────────────────────────────────────
export type LiquidityLevel = 'Very High' | 'High' | 'Good' | 'Moderate' | 'Moderate-Good';

export interface StockEntry {
  name: string;
  label: string;
  liquidity: LiquidityLevel;
  volatility: string;
  indices: string;
  why: string;
}

export interface LiquidityGroup {
  liquidity: string;
  stocks: StockEntry[];
}

// ─── Per-contract scan report ──────────────────────────────────────────────
export interface ContractScanEntry {
  underlying: string;
  symbol: string;
  strike: number;
  option_type: 'CE' | 'PE';
  expiry: string;
  moneyness: string;
  bars: number;
  premium_close: number;
  fired: boolean;
  fired_at_ms: number;
  reason: string;
}

export interface ScanReportSummary {
  generated_ms: number;
  scan_source: string;
  indices: string[];
  total_contracts: number;
  charted: number;
  fired: number;
  no_data: number;
  min_bars: number;
  max_bars: number;
  total_ce: number;
  total_pe: number;
  fired_ce: number;
  fired_pe: number;
}

export interface ScanReportResponse {
  summary: ScanReportSummary;
  entries: ContractScanEntry[];
}

// ─── Engine open positions ────────────────────────────────────────────────────
export type EngineVehicle = 'otm_options' | 'deep_itm_options' | 'futures';

export interface EngineOpenPosition {
  symbol: string;
  exchange: string;
  token: number;
  qty: number;
  lot_size: number;
  entry_premium: number;
  fill_price: number;
  stop_premium: number;
  status: string;
  direction: 'long' | 'short';
  vehicle: EngineVehicle;
  underlying: string;
  opened_ms: number;
  exit_reason: string;
  order_id: string;
}

export interface OpenPositionsResponse {
  positions: EngineOpenPosition[];
}

File: backend/app/api/v1/endpoints/kite_engine.py
"""Kite triple-SuperTrend engine endpoints — `/api/v1/kite/engine/*`.

Scoped to the calling Kite user. Advisory by default; auto-execute is opt-in via
config and runs through the same Kite order path + live-safety gate as manual
orders. Imports no other engine's strategy/signal/derivative logic.
"""
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException

from app.core.auth import UserContext, get_current_user
from app.core.logging import get_logger
from app.engines.triple_supertrend.config import TripleSupertrendConfig
from app.engines.triple_supertrend.schemas import (
    ActivityResponse, BacktestRequest, BacktestResponse, ContractScanEntry,
    EngineConfigModel, EngineDetailResponse, EngineOrderRequest, EngineOrderResponse,
    OpenPositionRecord, OpenPositionsResponse,
    ScanReportResponse, ScanReportSummary, SetupChart, SignalsResponse,
)
from app.services import live_safety
from app.services.exchanges.kite import accounts as kite_accounts
from app.services.exchanges.kite.errors import KiteError
from app.services.kite_engine import positions as kite_positions, service, state
from app.services.kite_engine.detail import build_detail
from app.services.kite_engine.market_hours import is_market_open
from app.services.kite_engine.scanner import build_setup_chart, scanner

log = get_logger(__name__)
router = APIRouter(prefix="/kite/engine", tags=["kite-engine"])


def _ts_cfg(c: EngineConfigModel) -> TripleSupertrendConfig:
    return TripleSupertrendConfig(trail_target=c.trail_target, early_lock=c.early_lock)


async def _client(user: UserContext):
    acct = kite_accounts.get_active(user.user_id)
    if not acct:
        raise HTTPException(409, "No active Kite account — add credentials and log in first.")
    return await kite_accounts.acquire_client(acct)   # warm, cached (no per-call close)


@router.get("/config", response_model=EngineConfigModel)
async def get_config(user: UserContext = Depends(get_current_user)) -> EngineConfigModel:
    return state.get_config(user.user_id)


@router.post("/config", response_model=EngineConfigModel)
async def set_config(body: EngineConfigModel,
                     user: UserContext = Depends(get_current_user)) -> EngineConfigModel:
    return state.set_config(user.user_id, body)


@router.post("/config/reset", response_model=EngineConfigModel)
async def reset_config(user: UserContext = Depends(get_current_user)) -> EngineConfigModel:
    return state.set_config(user.user_id, EngineConfigModel())


@router.post("/backtest", response_model=BacktestResponse)
async def backtest(body: BacktestRequest,
                   user: UserContext = Depends(get_current_user)) -> BacktestResponse:
    """Honest options backtest (workstream H). data_mode synthetic | real | both:
    synthetic replays the signal on real underlying history with BS-modeled premium
    (full history, modeled price); real replays an actual live-contract premium
    series (true price, short lookback); both runs each + reports BS-vs-real drift.
    Read-only: no orders, no live-safety gate."""
    from app.services.kite_engine import backtest_service
    client = await _client(user)
    try:
        result = await backtest_service.run_backtest(client, body)
    except KiteError as exc:
        raise HTTPException(502, f"Kite data fetch failed: {exc}")
    return BacktestResponse(**result)


@router.get("/signals", response_model=SignalsResponse)
async def signals(user: UserContext = Depends(get_current_user)) -> SignalsResponse:
    uid = user.user_id
    us = scanner.snapshot(uid)
    st = state.status(uid)
    return SignalsResponse(generated_ms=us.generated_ms, scanning=us.scanning, scanning_label=us.scanning_label, rows=us.rows,
                           next_scan_ms=st.next_scan_ms, auto_scan=service.is_auto_running(),
                           market_open=is_market_open())


@router.get("/activity", response_model=ActivityResponse)
async def activity(limit: int = 2000,
                   user: UserContext = Depends(get_current_user)) -> ActivityResponse:
    uid = user.user_id
    st = state.status(uid)
    us = scanner.snapshot(uid)
    return ActivityResponse(
        events=state.activity(uid, limit), scanning=st.scanning, auto_scan=service.is_auto_running(),
        last_scan_ms=st.last_scan_ms, next_scan_ms=st.next_scan_ms, signal_count=st.signal_count,
        scanning_label=us.scanning_label if us.scanning else "",
    )


@router.get("/server-logs")
async def server_logs(limit: int = 300,
                      user: UserContext = Depends(get_current_user)) -> dict:
    """Recent backend server logs (in-memory ring buffer) for the Kite Terminal.
    Lets the UI interleave real server logs with engine activity. Read-only."""
    from app.core.logging import recent_logs
    return {"logs": recent_logs(limit)}


@router.post("/scan", response_model=SignalsResponse)
async def run_scan(user: UserContext = Depends(get_current_user)) -> SignalsResponse:
    """Manual scan trigger (the background loop also scans automatically)."""
    uid = user.user_id
    client = await _client(user)
    await service.scan_user(client, uid)
    us = scanner.snapshot(uid)
    st = state.status(uid)
    return SignalsResponse(generated_ms=us.generated_ms, scanning=us.scanning, scanning_label=us.scanning_label, rows=us.rows,
                           next_scan_ms=st.next_scan_ms, auto_scan=service.is_auto_running(),
                           market_open=is_market_open())


@router.post("/scan/cancel", response_model=SignalsResponse)
async def cancel_scan(user: UserContext = Depends(get_current_user)) -> SignalsResponse:
    """Force-stop a running scan. Returns the current snapshot after cancellation."""
    uid = user.user_id
    cancelled = scanner.cancel(uid)
    if cancelled:
        state.log(uid, "info", "Scan cancelled by user.")
        state.set_scanning(uid, False)
        state.set_cooldown(uid)
    us = scanner.snapshot(uid)
    st = state.status(uid)
    return SignalsResponse(generated_ms=us.generated_ms, scanning=us.scanning, scanning_label=us.scanning_label, rows=us.rows,
                           next_scan_ms=st.next_scan_ms, auto_scan=service.is_auto_running(),
                           market_open=is_market_open())


@router.get("/setup/{token}", response_model=SetupChart)
async def setup(token: int, underlying: str = "",
                user: UserContext = Depends(get_current_user)) -> SetupChart:
    client = await _client(user)
    return await build_setup_chart(client, token, underlying, _ts_cfg(state.get_config(user.user_id)))


@router.post("/order", response_model=EngineOrderResponse)
async def place_order(body: EngineOrderRequest,
                     user: UserContext = Depends(get_current_user)) -> EngineOrderResponse:
    """Place a manual BUY/SELL from the detail panel — same live-safety gate as the
    standard order path, but logged to the engine terminal with full error surfacing."""
    res = await service.place_manual_order(
        user.user_id, body.option_symbol, body.side, body.quantity, body.exchange)
    if res["status"] == "blocked":
        raise HTTPException(423, detail={"reason": res.get("reason"), "code": res.get("code")})
    if res["status"] == "error":
        raise HTTPException(502, detail=res.get("message", "Order failed"))
    return EngineOrderResponse(
        order_id=res.get("order_id", ""), status=res["status"],
        message=res.get("message", ""))


@router.get("/detail/{token}", response_model=EngineDetailResponse)
async def detail(token: int, timestamp_ms: int = 0, user: UserContext = Depends(get_current_user)) -> EngineDetailResponse:
    """Trigger context + live underlying price + per-leg quote/depth/greeks for a
    ready signal (BUY/SELL are placed via the standard /kite/orders endpoint)."""
    client = await _client(user)
    d = await build_detail(client, user.user_id, token, timestamp_ms)
    if d is None:
        raise HTTPException(404, "No ready signal for that instrument in the latest scan.")
    return d


@router.get("/scan-report", response_model=ScanReportResponse)
async def scan_report(user: UserContext = Depends(get_current_user)) -> ScanReportResponse:
    """Per-contract scan trace — every option contract evaluated, with bars, premium,
    and reason. Shows exactly which contracts fired and why others didn't."""
    uid = user.user_id
    snap = scanner.snapshot(uid)
    diag = snap.diag
    entries = [
        ContractScanEntry(
            underlying=c.underlying, symbol=c.symbol, strike=c.strike,
            option_type=c.option_type, expiry=c.expiry, moneyness=c.moneyness,
            bars=c.bars, premium_close=c.premium_close, fired=c.fired,
            fired_at_ms=c.fired_at_ms, reason=c.reason,
        )
        for c in diag.contracts
    ]
    total_ce = sum(1 for c in diag.contracts if c.option_type == "CE")
    total_pe = sum(1 for c in diag.contracts if c.option_type == "PE")
    fired_ce = sum(1 for c in diag.contracts if c.option_type == "CE" and c.fired)
    fired_pe = sum(1 for c in diag.contracts if c.option_type == "PE" and c.fired)
    summary = ScanReportSummary(
        generated_ms=snap.generated_ms,
        scan_source="derivatives",  # contracts = derivatives only
        indices=[],  # TODO: pull from config if needed
        total_contracts=len(entries),
        charted=diag.deriv_charts,
        fired=diag.deriv_fired,
        no_data=diag.deriv_no_data,
        min_bars=diag.deriv_min_bars,
        max_bars=diag.deriv_max_bars,
        total_ce=total_ce,
        total_pe=total_pe,
        fired_ce=fired_ce,
        fired_pe=fired_pe,
    )
    return ScanReportResponse(summary=summary, entries=entries)


@router.get("/open-positions", response_model=OpenPositionsResponse)
async def open_positions(user: UserContext = Depends(get_current_user)) -> OpenPositionsResponse:
    """Return the engine's currently tracked open positions including vehicle/direction labels."""
    uid = user.user_id
    records = [
        OpenPositionRecord(
            symbol=p.symbol, exchange=p.exchange, token=p.token,
            qty=p.qty, lot_size=p.lot_size,
            entry_premium=p.entry_premium, fill_price=p.fill_price,
            stop_premium=p.stop_premium, status=p.status,
            direction=p.direction, vehicle=p.vehicle, underlying=p.underlying,
            opened_ms=p.opened_ms, exit_reason=p.exit_reason, order_id=p.order_id,
        )
        for p in kite_positions.open_positions(uid)
    ]
    return OpenPositionsResponse(positions=records)


@router.delete("/open-positions/{symbol}", response_model=OpenPositionsResponse)
async def close_position(symbol: str, user: UserContext = Depends(get_current_user)) -> OpenPositionsResponse:
    """Manually close (mark-closed) a tracked position without placing a broker order.
    Cancels any live broker GTT stop before closing, so it can't fire after removal.
    Use when an order was filled outside the engine or to clean up stale entries."""
    uid = user.user_id
    from app.services.kite_engine import protective_stop as pstop
    from app.services.exchanges.kite import ticker_manager
    p = kite_positions.get(uid, symbol)
    if p:
        try:
            client = await _client(user)
            if p.gtt_id:
                await pstop.cancel_stop(client, p.gtt_id)
            if p.token:
                await ticker_manager.unsubscribe(uid, [p.token])
        except Exception:  # noqa: BLE001
            pass
    kite_positions.close(uid, symbol, reason="manual_close")
    records = [
        OpenPositionRecord(
            symbol=p.symbol, exchange=p.exchange, token=p.token,
            qty=p.qty, lot_size=p.lot_size,
            entry_premium=p.entry_premium, fill_price=p.fill_price,
            stop_premium=p.stop_premium, status=p.status,
            direction=p.direction, vehicle=p.vehicle, underlying=p.underlying,
            opened_ms=p.opened_ms, exit_reason=p.exit_reason, order_id=p.order_id,
        )
        for p in kite_positions.open_positions(uid)
    ]
    return OpenPositionsResponse(positions=records)


@router.get("/stock-registry")
async def stock_registry() -> list[dict]:
    """Return the curated stock registry with liquidity / volatility metadata,
    plus a separate optional-stocks group for the '+' picker."""
    from app.services.kite_engine.stock_registry import OPTIONAL_STOCKS, STOCK_REGISTRY, STOCKS_BY_LIQUIDITY
    groups = [
        {"liquidity": liq, "stocks": [e.to_dict() for e in entries]}
        for liq in ["Very High", "High", "Good"]
        if liq in STOCKS_BY_LIQUIDITY
        for entries in [STOCKS_BY_LIQUIDITY[liq]]
    ]
    groups.append({"liquidity": "optional", "stocks": [e.to_dict() for e in OPTIONAL_STOCKS]})
    return groups
